### PR TITLE
[Swift in WebKit] Make it possible to access internal project headers within Swift

### PR DESCRIPTION
**Pull request by rr-codes**
#### 5a154498add89a1c3d786cfc8e99d500b4daa237
<pre>
[Swift in WebKit] Make it possible to access internal project headers within Swift
<a href="https://bugs.webkit.org/show_bug.cgi?id=282904">https://bugs.webkit.org/show_bug.cgi?id=282904</a>
<a href="https://rdar.apple.com/139610042">rdar://139610042</a>

Reviewed by Elliott Williams.

Add an internal Swift module so that Swift files can access internal WebKit project headers.

* Source/WebKit/Configurations/WebKit.xcconfig:
* Source/WebKit/Modules/Internal/WebKitInternal.h: Added.
* Source/WebKit/Modules/Internal/module.modulemap: Added.
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/286465@main">https://commits.webkit.org/286465@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/f588f274e058e91fb79a6c5be5ce17dc50b50391

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/76070 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/55099 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/28969 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/80570 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/27336 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/64241 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/3395 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/59641 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/27336 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/79137 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/49531 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/65325 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39998 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/46930 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/25663 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/68046 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/23147 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/82031 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/3439 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/2207 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/67871 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/3593 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/65293 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/67182 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16739 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11132 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/9252 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/3389 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/6195 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/3410 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/6807 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/4261 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | |

### CODE REVIEW

File: Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h [hunk 1]
```diff
@@ -494,3 +498,9 @@ RetainPtr<NSError> nsErrorFromExceptionDetails(const WebCore::ExceptionDetails&)
 - (void)_requestTextExtractionForSwift:(WKTextExtractionRequest *)context;
 - (void)_requestTextExtraction:(CGRect)rect completionHandler:(void(^)(WKTextExtractionItem *))completionHandler;
 @end
+
+#endif // __cplusplus
+
+@interface WKWebView (NonCpp)
+
+@end
```
Comment (emw-apple) [inline Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:506]: Do we really want to re-declare ObjC interfaces from above? I would definitely prefer a more fine-grained use of `#ifdef __cplusplus` if you think that is possible.

File: Source/WebKit/Modules/Internal/module.modulemap [hunk 2]
```diff
@@ -0,0 +1,4 @@
+module WebKit_Internal [system] {
```
Comment (emw-apple) [inline Source/WebKit/Modules/Internal/module.modulemap]: per https://clang.llvm.org/docs/Modules.html#module-maps:
> The system attribute specifies that the module is a system module. When a system module is rebuilt, all of the module’s headers will be considered system headers, which suppresses warnings. This is equivalent to placing #pragma GCC system_header in each of the module’s headers.

I would think we do _not_ want these semantics, since this module is never used by clients of WebKit.

File: Source/WebKit/Modules/Internal/module.modulemap
Comment (emw-apple) [inline Source/WebKit/Modules/Internal/module.modulemap:1]: Worth noting that for our public and private modules, we are set up to preprocess their respective modulemaps, so that we can use things like `#if ENABLE(...)` checks from WTF. I don't see a reason why we need to do that for the internal module, though.

File: Source/WebKit/Configurations/WebKit.xcconfig [hunk 3]
```diff
@@ -62,6 +62,8 @@ BUILD_LIBRARY_FOR_DISTRIBUTION = YES;
 SWIFT_INSTALL_OBJC_HEADER = NO;
 SWIFT_LIBRARY_LEVEL = api;
 
+USER_HEADER_SEARCH_PATHS = $(SRCROOT)/Modules/Internal;
```
Comment (emw-apple) [inline Source/WebKit/Configurations/WebKit.xcconfig]: Does it work to pass this in `SWIFT_INCLUDE_PATHS` instead?

File: Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h (hunk 1 above)
Comment (rr-codes) [reply Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:506]: I originally had the same thought, but it seems like well over half of the file relies on C++ stuff, so I figured this was easier to follow than adding a ton of `#ifdef __cplusplus` everywhere. I can change it though if you want!

File: Source/WebKit/WebKit.xcodeproj/project.pbxproj [hunk 4]
```diff
@@ -16780,6 +16794,7 @@
 				950FECF5285027200002DE4E /* PaymentTokenContext.h in Headers */,
 				C1E123BA20A11573002646F4 /* PDFContextMenu.h in Headers */,
 				33AE61252B20F13B00E1C215 /* PDFKitSPI.h in Headers */,
+				07FB8E162CE1438D00B09C3C /* PDFPluginIdentifier.h in Headers */,
```
Comment (emw-apple) [inline Source/WebKit/WebKit.xcodeproj/project.pbxproj]: Seems unrelated?

File: Source/WebKit/WebKit.xcodeproj/project.pbxproj [hunk 5]
```diff
@@ -163,6 +164,7 @@
 		07E4BDC72A3A7089000D5509 /* _WKWebViewTextInputNotifications.h in Headers */ = {isa = PBXBuildFile; fileRef = 07E4BDC52A3A7089000D5509 /* _WKWebViewTextInputNotifications.h */; };
 		07E4BDC82A3A7089000D5509 /* _WKWebViewTextInputNotifications.mm in Sources */ = {isa = PBXBuildFile; fileRef = 07E4BDC62A3A7089000D5509 /* _WKWebViewTextInputNotifications.mm */; };
 		07F327F22CB09B74006D9918 /* _WKTextPreview.h in Headers */ = {isa = PBXBuildFile; fileRef = 07F327F02CB085A4006D9918 /* _WKTextPreview.h */; settings = {ATTRIBUTES = (Private, ); }; };
+		07FB8E162CE1438D00B09C3C /* PDFPluginIdentifier.h in Headers */ = {isa = PBXBuildFile; fileRef = DF74275C24F4955000F8ABE9 /* PDFPluginIdentifier.h */; };
```
Comment (emw-apple) [inline Source/WebKit/WebKit.xcodeproj/project.pbxproj]: Seems unrelated?

**Review by emw-apple — APPROVED** [review]
LGTM, but please remove these two pbxproj changes:

File: Source/WebKit/WebKit.xcodeproj/project.pbxproj (hunk 5 above)
Comment (rr-codes) [reply Source/WebKit/WebKit.xcodeproj/project.pbxproj]: oh wow, thanks for noticing!
